server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -1236,4 +1251,457 @@ synchronized ClusterStateUpdateStats getStatistics() {
+            @Override
+            public void onRejection(FailedToCommitClusterStateException e) {
+                final var items = queueSize.getAndSet(0);
+                for (int i = 0; i < items; i++) {
+                    final var entry = queue.poll();
+                    assert entry != null;
+                    entry.onRejection(e);
+                }
+            }
+
+            @Override
+            public void run() {
+                assert executing.isEmpty() : executing;
+                final var entryCount = queueSize.getAndSet(0);
+                var taskCount = 0;
+                for (int i = 0; i < entryCount; i++) {
+                    final var entry = queue.poll();
+                    assert entry != null;
+                    if (entry.acquireForExecution()) {
+                        taskCount += 1;
+                        executing.add(entry);
+                    }
+                }
+                if (taskCount == 0) {
+                    return;
+                }
+                final var finalTaskCount = taskCount;
+                final var tasks = new ArrayList<ExecutionResult<T>>(finalTaskCount);
+                for (final var entry : executing) {
+                    tasks.add(
+                        new ExecutionResult<>(entry.source(), entry.task(), threadPool.getThreadContext(), entry.storedContextSupplier())
+                    );
+                }
+                try {
+                    batchConsumer.runBatch(executor, tasks, new BatchSummary(() -> buildTasksDescription(tasks)));
+                } finally {
+                    assert executing.size() == finalTaskCount;
+                    executing.clear();
+                }
+            }
+
+            private String buildTasksDescription(List<ExecutionResult<T>> tasks) {
+                final var tasksBySource = new HashMap<String, List<T>>();
+                for (final var entry : tasks) {
+                    tasksBySource.computeIfAbsent(entry.getSource(), ignored -> new ArrayList<>()).add(entry.getTask());
+                }
+
+                final var output = new StringBuilder();
+                Strings.collectionToDelimitedStringWithLimit((Iterable<String>) () -> tasksBySource.entrySet().stream().map(entry -> {
+                    var tasksDescription = executor.describeTasks(entry.getValue());
+                    return tasksDescription.isEmpty() ? entry.getKey() : entry.getKey() + "[" + tasksDescription + "]";
+                }).filter(s -> s.isEmpty() == false).iterator(), ", ", "", "", MAX_TASK_DESCRIPTION_CHARS, output);
+                if (output.length() > MAX_TASK_DESCRIPTION_CHARS) {
+                    output.append(" (").append(tasks.size()).append(" tasks in total)");
+                }
+                return output.toString();
+            }
+
+            @Override
+            public Stream<PendingClusterTask> getPending(long currentTimeMillis) {
+                return Stream.concat(
```

Comment:
We remove from `queue` and add to `executing` above, leaving a few opportunities for tasks to be missed in the pending list:

1. There is a short window of time where it is in neither queue.
2. The consistency of the `stream` returned is weak, and we risk the executing set not representing everything added while the queue set may be empty.

It feels like a double buffering solution could fit, swapping `executing` and `queue`, though there is some trouble getting it to be lock-free.

Reply:
Yes I struggled to get better consistency in this area. I think this is not materially worse than what we have today, if you look at how `PrioritizedEsThreadPoolExecutor` manages the `current` collection.

Reply:
Got it, thanks, I think I agree to that it is sort of the same consistency, so good by me.